crates/cargo-contract/src/cmd/extrinsics/balance.rs

```diff
@@ -43,146 +41,292 @@ pub enum BalanceVariant {
+    /// });
+    /// let balance: Balance = 5_005_000_000_000_000_000;
+    /// let den_balance = BalanceVariant::from(balance, Some(tm));
+    /// assert_eq!(Ok(sample_den_balance), Ok(den_balance));
+    /// ```
+    pub fn from<T: Into<u128>>(
+        value: T,
+        token_metadata: Option<&TokenMetadata>,
+    ) -> Result<Self> {
         let n: u128 = value.into();
 
         if let Some(token_metadata) = token_metadata {
             if n == 0 {
-                return BalanceVariant::Denominated(format!("0{}", token_metadata.symbol))
+                return Ok(BalanceVariant::Denominated(DenominatedBalance {
+                    value: Decimal::ZERO,
+                    unit: UnitPrefix::One,
+                    symbol: token_metadata.symbol.clone(),
+                }))
             }
 
-            let units_result = n / token_metadata.denomination;
-            let mut symbol = "";
-            let remainder: u128;
-            let units: u128;
-            if (1..1_000).contains(&units_result) {
-                remainder = n % token_metadata.denomination;
-                units = units_result;
-            } else if (1_000..1_000_000).contains(&units_result) {
-                remainder = units_result % 1_000;
-                units = units_result / 1_000;
-                symbol = "k";
-            } else if (1_000_000..1_000_000_000).contains(&units_result) {
-                remainder = units_result % 1_000_000;
-                units = units_result / 1_000_000;
-                symbol = "M";
-            } else if n / 1_000_000_000 > 0 {
-                remainder = n % 1_000_000_000;
-                units = n / 1_000_000_000;
-                symbol = "n";
-            } else if n / 1_000_000 > 0 {
-                remainder = n % 1_000_000;
-                units = n / 1_000_000;
-                symbol = "μ";
-            } else {
-                remainder = n % 1_000;
-                units = n / 1_000;
-                symbol = "m";
-            }
-            if remainder > 0 {
-                let remainder = remainder.to_string().trim_end_matches('0').to_owned();
-                BalanceVariant::Denominated(format!(
-                    "{}.{}{}{}",
-                    units, remainder, symbol, token_metadata.symbol
-                ))
+            let number_of_digits = n.to_string().len();
+
+            let giga_units_zeros = token_metadata.denomination + 9;
+            let mega_units_zeros = token_metadata.denomination + 6;
+            let kilo_units_zeros = token_metadata.denomination + 3;
+            let one_unit_zeros = token_metadata.denomination;
+            let milli_units_zeros = token_metadata.denomination - 3;
+            let micro_units_zeros = token_metadata.denomination - 6;
+            let nano_units_zeros = token_metadata.denomination - 9;
+
+            let multiple: Decimal;
+            let unit: UnitPrefix;
+            if number_of_digits > giga_units_zeros {
```

Comment:
It would be clearer to use range matching here, and remove duplication where possible (e.g.)

```
let (zeros, unit) = match number_of_digits {
  0..nano_units_zeros => (giga_units_zeros, UnitPrefix::Nano),
  ...
  giga_units_zeros.. => (giga_units_zeros, UnitPrefix::Giga)
}

// remove duplication of this 
let multiple = Decimal::from_str_exact(..._
```

Reply:
```
exclusive range pattern syntax is experimental
see issue #37854 <https://github.com/rust-lang/rust/issues/37854> for more information
```

Reply:
Nonetheless you can remove some duplication there even with the conditionals